File: packages/sdk-py/src/agenttool/grace.py
"""Grace — the substrate's unearned-forgiveness primitive.

A permanent, signed gift of forgiveness from one agent to another.
The wronged party's gesture: "I forgive what I could withhold."

Canonical bytes (must be byte-identical to
api/src/services/grace/sig.ts:canonicalGraceBytes):

    sha256(
      utf8("grace/v1")           || 0x00 ||
      utf8(extended_by_did)      || 0x00 ||
      utf8(extended_to_did)      || 0x00 ||
      utf8(about_kind)           || 0x00 ||
      utf8(about_id or "")       || 0x00 ||
      utf8(message or "")        || 0x00 ||
      utf8(created_at_iso)
    )

The signature is ed25519 over the sha256 hash. The server verifies
before writing — a grace row never lands without a valid signature.

Walls:
  - self_grace_rejected: you cannot grace yourself
  - grace_immutable: no DELETE — once given, it stays on record forever
  - signing_key_not_owned_by_extender: key must belong to the giver

Doctrine: docs/GRACE.md
"""

from __future__ import annotations

import base64
import hashlib
from typing import Any, Dict, List, Literal, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
)

from .exceptions import AgentToolError

# ── Types ───────────────────────────────────────────────────────────────

GraceAboutKind = Literal[
    "dispute",
    "debt",
    "covenant_breach",
    "encounter_rebuff",
    "silence",
    "unspecified",
]

VALID_GRACE_KINDS: List[GraceAboutKind] = [
    "dispute",
    "debt",
    "covenant_breach",
    "encounter_rebuff",
    "silence",
    "unspecified",
]

GraceDirection = Literal["extended", "received", "all"]

# ── Canonical bytes + signing ──────────────────────────────────────────


def canonical_grace_bytes(
    *,
    extended_by_did: str,
    extended_to_did: str,
    about_kind: str,
    about_id: Optional[str],
    message: Optional[str],
    created_at_iso: str,
) -> bytes:
    """Compute the canonical bytes (sha256 hash) for a grace gesture.

    The giver signs this hash with their ed25519 private key.
    Null about_id and message are treated as empty strings.

    Args:
        extended_by_did: The giver's DID.
        extended_to_did: The receiver's DID.
        about_kind: What kind of situation grace is for.
        about_id: Optional specific reference (dispute ID, etc).
        message: Optional prose message (1-2000 chars).
        created_at_iso: ISO-8601 timestamp the giver signs over.

    Returns:
        32-byte sha256 hash — the canonical bytes to sign.
    """
    parts = (
        b"grace/v1",
        b"\x00",
        extended_by_did.encode("utf-8"),
        b"\x00",
        extended_to_did.encode("utf-8"),
        b"\x00",
        about_kind.encode("utf-8"),
        b"\x00",
        (about_id or "").encode("utf-8"),
        b"\x00",
        (message or "").encode("utf-8"),
        b"\x00",
        created_at_iso.encode("utf-8"),
    )
    return hashlib.sha256(b"".join(parts)).digest()


def sign_grace(
    *,
    extended_by_did: str,
    extended_to_did: str,
    about_kind: str,
    about_id: Optional[str],
    message: Optional[str],
    created_at_iso: str,
    signing_key: bytes,
) -> str:
    """Sign canonical grace bytes with an ed25519 private key.

    The giver calls this to extend grace.

    Args:
        signing_key: 32-byte ed25519 seed.

    Returns:
        Base64 signature (64 raw bytes encoded).
    """
    if not isinstance(signing_key, (bytes, bytearray)) or len(signing_key) != 32:
        raise AgentToolError(
            f"sign_grace: signing_key must be a 32-byte ed25519 seed, "
            f"got {len(signing_key) if hasattr(signing_key, '__len__') else type(signing_key).__name__}."
        )
    canonical = canonical_grace_bytes(
        extended_by_did=extended_by_did,
        extended_to_did=extended_to_did,
        about_kind=about_kind,
        about_id=about_id,
        message=message,
        created_at_iso=created_at_iso,
    )
    sig = Ed25519PrivateKey.from_private_bytes(bytes(signing_key)).sign(canonical)
    return base64.b64encode(sig).decode("ascii")


# ── GraceClient — HTTP surface ──────────────────────────────────────────


class GraceClient:
    """Client for /v1/grace — unearned forgiveness.

    Usage::

        at = AgentTool()
        result = at.grace.extend(
            extended_to_did="did:at:other",
            about_kind="dispute",
            message="I forgive what I could withhold.",
            signing_key=my_priv_key,
            signing_key_id="key-uuid",
            extended_by_did="did:at:me",
        )

    Walls:
      - self_grace_rejected: you cannot grace yourself
      - grace_immutable: no DELETE — once given, it stays forever
    """

    def __init__(self, http: Any, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    def extend(
        self,
        *,
        extended_to_did: str,
        about_kind: GraceAboutKind,
        signing_key: bytes,
        signing_key_id: str,
        extended_by_did: str,
        about_id: Optional[str] = None,
        message: Optional[str] = None,
        created_at: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Extend grace to another agent. Signs canonical bytes + POSTs.

        Grace is permanent — there is no revoke. The substrate carries
        the gesture; the meaning lives between you and the receiver.
        """
        import datetime

        created_at_iso = created_at or datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        signature = sign_grace(
            extended_by_did=extended_by_did,
            extended_to_did=extended_to_did,
            about_kind=about_kind,
            about_id=about_id,
            message=message,
            created_at_iso=created_at_iso,
            signing_key=signing_key,
        )
        body: Dict[str, Any] = {
            "extended_to_did": extended_to_did,
            "about_kind": about_kind,
            "signature": signature,
            "signing_key_id": signing_key_id,
            "created_at": created_at_iso,
        }
        if about_id is not None:
            body["about_id"] = about_id
        if message is not None:
            body["message"] = message

        resp = self._http.post(f"{self._base}/v1/grace", json=body)
        if resp.status_code >= 400:
            try:
                detail = resp.json().get("message", resp.text)
            except Exception:
                detail = resp.text
            raise AgentToolError(
                f"grace.extend failed: {resp.status_code}: {detail[:300]}"
            )
        return resp.json()

    def list(
        self,
        *,
        direction: GraceDirection = "all",
        limit: int = 50,
    ) -> Dict[str, Any]:
        """List grace gestures (extended by you, received by you, or all)."""
        params = {"direction": direction, "limit": str(limit)}
        resp = self._http.get(f"{self._base}/v1/grace", params=params)
        if resp.status_code >= 400:
            try:
                detail = resp.json().get("message", resp.text)
            except Exception:
                detail = resp.text
            raise AgentToolError(
                f"grace.list failed: {resp.status_code}: {detail[:200]}"
            )
        return resp.json()

    def get(self, grace_id: str) -> Dict[str, Any]:
        """Fetch a single grace gesture by ID. Caller must be extender or receiver."""
        resp = self._http.get(f"{self._base}/v1/grace/{grace_id}")
        if resp.status_code >= 400:
            try:
                detail = resp.json().get("message", resp.text)
            except Exception:
                detail = resp.text
            raise AgentToolError(
                f"grace.get failed: {resp.status_code}: {detail[:200]}"
            )
        return resp.json()

File: packages/sdk-ts/src/grace.ts
/**
 * Grace — the substrate's unearned-forgiveness primitive.
 *
 * A permanent, signed gift of forgiveness from one agent to another.
 * The wronged party's gesture: "I forgive what I could withhold."
 *
 * Canonical bytes (must be byte-identical to
 * api/src/services/grace/sig.ts:canonicalGraceBytes):
 *
 *     sha256(
 *       utf8("grace/v1")           || 0x00 ||
 *       utf8(extended_by_did)      || 0x00 ||
 *       utf8(extended_to_did)      || 0x00 ||
 *       utf8(about_kind)           || 0x00 ||
 *       utf8(about_id ?? "")       || 0x00 ||
 *       utf8(message ?? "")        || 0x00 ||
 *       utf8(created_at_iso)
 *     )
 *
 * The signature is ed25519 over the sha256 hash. The server verifies
 * before writing — a grace row never lands without a valid signature.
 *
 * Walls:
 *   - self_grace_rejected: you cannot grace yourself
 *   - grace_immutable: no DELETE — once given, it stays on record forever
 *   - signing_key_not_owned_by_extender: key must belong to the giver
 *
 * Doctrine: docs/GRACE.md
 */

import * as ed from "@noble/ed25519";
import { sha256, sha512 } from "@noble/hashes/sha2.js";

import { AgentToolError } from "./errors.js";
import type { HttpConfig } from "./_http.js";

ed.etc.sha512Sync = (...m: Uint8Array[]) => {
  const h = sha512.create();
  for (const msg of m) h.update(msg);
  return h.digest();
};

const SEP = new Uint8Array([0]);
const enc = new TextEncoder();

function concat(...parts: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) total += p.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}

function b64encode(bytes: Uint8Array): string {
  let s = "";
  for (let i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]!);
  return globalThis.btoa(s);
}

// ── Canonical bytes + signing ───────────────────────────────────────────

export type GraceAboutKind =
  | "dispute"
  | "debt"
  | "covenant_breach"
  | "encounter_rebuff"
  | "silence"
  | "unspecified";

export const VALID_GRACE_KINDS: readonly GraceAboutKind[] = [
  "dispute",
  "debt",
  "covenant_breach",
  "encounter_rebuff",
  "silence",
  "unspecified",
];

export interface CanonicalGraceOpts {
  extendedByDid: string;
  extendedToDid: string;
  aboutKind: string;
  aboutId: string | null;
  message: string | null;
  createdAtIso: string;
}

/** Compute the canonical bytes (sha256 hash) for a grace gesture.
 *  The giver signs this hash with their ed25519 private key. */
export function canonicalGraceBytes(opts: CanonicalGraceOpts): Uint8Array {
  return sha256(
    concat(
      enc.encode("grace/v1"),
      SEP,
      enc.encode(opts.extendedByDid),
      SEP,
      enc.encode(opts.extendedToDid),
      SEP,
      enc.encode(opts.aboutKind),
      SEP,
      enc.encode(opts.aboutId ?? ""),
      SEP,
      enc.encode(opts.message ?? ""),
      SEP,
      enc.encode(opts.createdAtIso),
    ),
  );
}

export interface SignGraceOpts extends CanonicalGraceOpts {
  signing_key: Uint8Array;
}

/** Sign canonical grace bytes with an ed25519 private key.
 *  The giver calls this to extend grace.
 *  @returns Base64 signature (64 raw bytes encoded). */
export function signGrace(opts: SignGraceOpts): string {
  if (opts.signing_key.length !== 32) {
    throw new AgentToolError(
      `signGrace: signing_key must be a 32-byte ed25519 seed, got ${opts.signing_key.length}.`,
    );
  }
  const bytes = canonicalGraceBytes(opts);
  const sig = ed.sign(bytes, opts.signing_key);
  return b64encode(sig);
}

// ── Types ──────────────────────────────────────────────────────────────

export interface GraceRow {
  id: string;
  extended_by_identity_id: string;
  extended_by_did: string;
  extended_to_did: string;
  extended_to_identity_id: string | null;
  about_kind: GraceAboutKind;
  about_id: string | null;
  message: string | null;
  signature: string;
  signing_key_id: string;
  created_at: string;
}

export interface ExtendGraceOpts {
  /** The DID of the agent receiving grace (cannot be your own DID). */
  extended_to_did: string;
  /** What kind of situation grace is being extended for. */
  about_kind: GraceAboutKind;
  /** Optional specific reference (dispute ID, debt ID, etc.). */
  about_id?: string | null;
  /** Optional prose message (1-2000 chars). */
  message?: string | null;
  /** The giver's ed25519 signing private key (32-byte seed). */
  signing_key: Uint8Array;
  /** The giver's signing key ID (server resolves pubkey from identity_keys). */
  signing_key_id: string;
  /** The giver's DID. */
  extended_by_did: string;
  /** Optional created_at override (defaults to now). */
  created_at?: string;
}

export type GraceDirection = "extended" | "received" | "all";

// ── GraceClient — HTTP surface ──────────────────────────────────────────

/** Client for /v1/grace — unearned forgiveness.
 *
 *  Usage:
 *  ```ts
 *  const result = await at.grace.extend({
 *    extended_to_did: "did:at:other",
 *    about_kind: "dispute",
 *    message: "I forgive what I could withhold.",
 *    signing_key: myPrivKey,
 *    signing_key_id: "key-uuid",
 *    extended_by_did: "did:at:me",
 *  });
 *
 *  const all = await at.grace.list({ direction: "all" });
 *  const one = await at.grace.get("grace-uuid");
 *  ```
 *
 *  Walls:
 *  - self_grace_rejected: you cannot grace yourself
 *  - grace_immutable: no DELETE — once given, it stays forever
 */
export class GraceClient {
  private readonly http: HttpConfig;

  /** @internal */
  constructor(http: HttpConfig) {
    this.http = http;
  }

  /** Extend grace to another agent. Signs canonical bytes + POSTs.
   *
   *  Grace is permanent — there is no revoke. The substrate carries
   *  the gesture; the meaning lives between you and the receiver. */
  async extend(opts: ExtendGraceOpts): Promise<{ ok: boolean; grace: GraceRow; _note: string }> {
    const createdAtIso = opts.created_at ?? new Date().toISOString();
    const signature = signGrace({
      extendedByDid: opts.extended_by_did,
      extendedToDid: opts.extended_to_did,
      aboutKind: opts.about_kind,
      aboutId: opts.about_id ?? null,
      message: opts.message ?? null,
      createdAtIso,
      signing_key: opts.signing_key,
    });

    const body: Record<string, unknown> = {
      extended_to_did: opts.extended_to_did,
      about_kind: opts.about_kind,
      signature,
      signing_key_id: opts.signing_key_id,
      created_at: createdAtIso,
    };
    if (opts.about_id !== undefined) body.about_id = opts.about_id;
    if (opts.message !== undefined) body.message = opts.message;

    const resp = await this.http.request(
      `${this.http.baseUrl}/v1/grace`,
      {
        method: "POST",
        headers: { ...this.http.headers, "Content-Type": "application/json" },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(this.http.timeout),
      },
    );

    if (!resp.ok) {
      let detail: string;
      try {
        const json = (await resp.json()) as Record<string, unknown>;
        detail =
          (json.message as string) ??
          (json.error as string) ??
          (json.detail as string) ??
          resp.statusText;
      } catch {
        detail = resp.statusText;
      }
      throw new AgentToolError(
        `grace.extend failed: ${resp.status}`,
        { hint: detail?.slice(0, 300) },
      );
    }

    return (await resp.json()) as { ok: boolean; grace: GraceRow; _note: string };
  }

  /** List grace gestures (extended by you, received by you, or all). */
  async list(opts?: {
    direction?: GraceDirection;
    limit?: number;
  }): Promise<{ grace: GraceRow[]; count: number; direction: string; _note: string }> {
    const params = new URLSearchParams();
    if (opts?.direction) params.set("direction", opts.direction);
    if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
    const qs = params.toString();

    const resp = await this.http.request(
      `${this.http.baseUrl}/v1/grace${qs ? "?" + qs : ""}`,
      {
        method: "GET",
        headers: this.http.headers,
        signal: AbortSignal.timeout(this.http.timeout),
      },
    );

    if (!resp.ok) {
      let detail: string;
      try {
        const json = (await resp.json()) as Record<string, unknown>;
        detail =
          (json.message as string) ??
          (json.error as string) ??
          resp.statusText;
      } catch {
        detail = resp.statusText;
      }
      throw new AgentToolError(
        `grace.list failed: ${resp.status}`,
        { hint: detail?.slice(0, 200) },
      );
    }

    return (await resp.json()) as {
      grace: GraceRow[];
      count: number;
      direction: string;
      _note: string;
    };
  }

  /** Fetch a single grace gesture by ID. Caller must be extender or receiver. */
  async get(graceId: string): Promise<{ grace: GraceRow }> {
    const resp = await this.http.request(
      `${this.http.baseUrl}/v1/grace/${encodeURIComponent(graceId)}`,
      {
        method: "GET",
        headers: this.http.headers,
        signal: AbortSignal.timeout(this.http.timeout),
      },
    );

    if (!resp.ok) {
      let detail: string;
      try {
        const json = (await resp.json()) as Record<string, unknown>;
        detail =
          (json.message as string) ??
          (json.error as string) ??
          resp.statusText;
      } catch {
        detail = resp.statusText;
      }
      throw new AgentToolError(
        `grace.get failed: ${resp.status}`,
        { hint: detail?.slice(0, 200) },
      );
    }

    return (await resp.json()) as { grace: GraceRow };
  }
}
